src/batched/KokkosBatched_Gemm_DblBuf_Impl.hpp

```diff
@@ -0,0 +1,535 @@
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+#ifndef __KOKKOSBATCHED_GEMM_DBLBUF_IMPL_HPP__
+#define __KOKKOSBATCHED_GEMM_DBLBUF_IMPL_HPP__
+
+#include "KokkosBatched_Util.hpp"
+
+namespace KokkosBatched {
+/********************* BEGIN functor-level routines *********************/
+///
+/// Serial Impl
+/// ===========
+
+///
+/// Implemented:
+/// NT/NT, T/NT, NT/T, T/T
+///
+/// Not yet immplemented (ConjTranspose):
+/// CT/NT, NT/CT, CT/CT
+///
+
+/********************* BEGIN functor-level routines *********************/
+/********************* END   functor-level routines *********************/
+
+namespace Impl {
+/********************* BEGIN non-functor-level routines *********************/
+template <class ArgTransA, class ArgTransB, class ArgBatchSzDim,
+          class HandleType, class ScalarType, class AViewType, class BViewType,
+          class CViewType, class ArgBoundsCheck, int TILE_M, int TILE_N,
+          int TILE_K>
+class BatchedDblBufGemm {
+ private:
+  HandleType *const __handle;
+  AViewType __A;
+  BViewType __B;
+  CViewType __C;
+  ScalarType __alpha, __beta;
+  ArgTransA __transA_tag;
+  ArgTransB __transB_tag;
+  ArgBatchSzDim __batch_layout_tag;
+  ArgBoundsCheck __bounds_check_tag;
+  int __c_batch_size, __c_m, __c_n;
+
+  using view_value_type      = typename CViewType::value_type;
+  using layout_type          = typename CViewType::array_layout;
+  using device_type          = typename CViewType::device_type;
+  using execution_space_type = typename device_type::execution_space;
+  using scratch_space_type =
+      typename execution_space_type::scratch_memory_space;
+  using view_type_2d_scratch =
+      Kokkos::View<view_value_type **, Kokkos::LayoutRight, scratch_space_type>;
```

Comment:
I am just curious. Is using LayoutRight for scratch space giving best performance for both LayoutLeft and LayoutRight matrices?

Reply:
Yes. Given the loops in the operator, this makes sense since the inner most loop iterates over the right-most dimension.